redash/query_runner/__init__.py

```diff
@@ -303,3 +334,46 @@ def guess_type_from_string(string_value):
         pass
 
     return TYPE_STRING
+
+
+def with_ssh_tunnel(query_runner, details):
+    def tunnel(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            try:
+                remote_host, remote_port = query_runner.host, query_runner.port
+            except NotImplementedError:
+                raise NotImplementedError(
+                    "SSH tunneling is not implemented for this query runner yet."
+                )
+
+            stack = ExitStack()
+            try:
+                bastion_address = (details["ssh_host"], details.get("ssh_port", 22))
+                remote_address = (remote_host, remote_port)
+                auth = {
+                    "ssh_username": details["ssh_username"],
+                    **settings.dynamic_settings.ssh_tunnel_auth(),
+                }
+                server = stack.enter_context(
+                    open_tunnel(
+                        bastion_address, remote_bind_address=remote_address, **auth
+                    )
+                )
+            except Exception as error:
+                raise type(error)("SSH tunnel: {}".format(str(error)))
+            else:
+                with stack:
+                    try:
+                        query_runner.host, query_runner.port = server.local_bind_address
+                        result = f(*args, **kwargs)
+                    finally:
+                        query_runner.host, query_runner.port = remote_host, remote_port
+
+                    return result
```

Comment:
Would it make sense to lose `else` here and remove an indent? I avoid `try...except...else` unless I catch the exception _without_ re-raising it.

Reply:
Yeah this makes no sense. I had originally used a different implementation where it made sense but didn't change the function's outline as I refactored. 93a97486.